project/BuildPlugin.scala

```diff
@@ -0,0 +1,359 @@
+import com.lucidchart.sbt.scalafmt.ScalafmtSbtPlugin.autoImport.Sbt
+import com.typesafe.tools.mima.plugin.MimaKeys
+
+object BuildPlugin extends AutoPlugin {
+  override def requires = sbt.plugins.JvmPlugin
+  override def trigger = allRequirements
+  val autoImport = BuildKeys
+
+  override def projectSettings: Seq[Def.Setting[_]] = BuildImplementation.projectSettings
+  override def buildSettings: Seq[Def.Setting[_]] = BuildImplementation.buildSettings
+  override def globalSettings: Seq[Def.Setting[_]] = BuildImplementation.globalSettings
+}
+
+object BuildKeys {
+  import sbt.{ file, File, Developer, url }
+  import BuildImplementation.{ BuildDefaults, BuildResolvers }
+
+  val baseVersion: String = "1.1.0-SNAPSHOT"
+  val internalPath: File = file("internal")
+  val bridgeScalaVersions: List[String] =
+    List(Dependencies.scala212, Dependencies.scala211, Dependencies.scala210)
+
+  val ZincGitHomepage: URL = url("https://github.com/sbt/zinc")
+  val ScalaCenterMaintainer: Developer =
+    Developer("jvican", "Jorge Vicente Cantero", "@jvican", url("https://github.com/jvican"))
+
+  // Ids that we will use to name our projects in build.sbt
+  val CompilerInterfaceId = "compiler-interface"
+  val CompilerBridgeId = "compiler-bridge"
+  val ZincApiInfoId = "zinc-apiinfo"
+
+  // Defines the constants for the alternative publishing
+  val ZincAlternativeCacheName = "alternative-local"
+  val ZincAlternativeCacheDir: File = file(sys.props("user.home") + "/.ivy2/zinc-alternative")
+
+  // Defines several settings that are exposed to the projects definition in build.sbt
+  private[this] val noPublishSettings: Seq[Def.Setting[_]] = BuildDefaults.noPublishSettings
+  def noPublish(p: Project): Project = p.settings(noPublishSettings)
+
+  val adaptOptionsForOldScalaVersions: Seq[Def.Setting[_]] =
+    List(Keys.scalacOptions := BuildDefaults.zincScalacOptionsRedefinition.value)
+  // Sets up mima settings for modules that have to be binary compatible with Zinc 1.0.0
+  val mimaSettings: Seq[Def.Setting[_]] =
+    List(MimaKeys.mimaPreviousArtifacts := BuildDefaults.zincPreviousArtifacts.value)
+
+  import sbt.{ TaskKey, taskKey }
+  val scriptedPublishAll = taskKey[Unit]("Publishes all the Zinc artifacts for scripted")
+  val cleanSbtBridge: TaskKey[Unit] = taskKey[Unit]("Cleans the sbt bridge.")
+  val zincPublishLocal: TaskKey[Unit] =
+    taskKey[Unit]("Publishes Zinc artifacts to a alternative local cache.")
+  val zincPublishLocalSettings: Seq[Def.Setting[_]] = List(
+    Keys.resolvers += BuildResolvers.AlternativeLocalResolver,
+    zincPublishLocal := BuildDefaults.zincPublishLocal.value,
+  )
+
+  val tearDownBenchmarkResources: TaskKey[Unit] = taskKey[Unit]("Remove benchmark resources.")
+  val benchmarksTestDir = sbt.IO.createTemporaryDirectory
+
+  val sourcesForAllScalaVersionsSetting: Seq[Def.Setting[_]] =
+    List(Keys.unmanagedSourceDirectories ++= BuildDefaults.handleScalaSpecificSources.value)
```

Comment:
Why the indirection? We already have a brand-new and obscure "sourcesForAllScalaVersionsSetting", why not define here what it does instead of deferring to `BuildDefaults.handleScalaSpecificSources` and then depending on build.sbt to make sure to call `inCompileAndTest`?

Reply:
Yup, we can move `unmanagedSourceDirectories` to the build.sbt.